### ISSUE TITLE
Memory Leak with procstat

### ISSUE THREAD

**Issue by mafinn** [influxdata/telegraf#6807]
### Relevant telegraf.conf:
```toml
[global_tags]
  sc = "daf"
  p = "32"
  custom_version = "1.x"
  os_type = "win2016s"

[agent]
  interval = "1s"
  round_interval = false
  metric_batch_size = 1000
  metric_buffer_limit = 5000
  collection_jitter = "1s"
  flush_interval = "1s"
  flush_jitter = "1s"
  precision = "s"
  debug = true
  quiet = false
  logfile = "/Program Files/Telegraf/telegraf.log"
  hostname = "win2016s"
  omit_hostname = false

[[outputs.influxdb]]
  urls = [ "http://1.1.1.1:8086" ]
  database = "telegraf"
  retention_policy = "24hours"
  precision = "m"

[[inputs.cpu]]
  percpu = true
  totalcpu = true
  collect_cpu_time = false
  report_active = false

[[inputs.mem]]  

[[inputs.procstat]]
  interval = "1s"
  exe = ".*"
  pid_finder = "native"

[[inputs.internal]]


```

### System info:
![image](https://user-images.githubusercontent.com/47539414/70962409-20082300-2053-11ea-8827-8b3f8f24d889.png)

![image](https://user-images.githubusercontent.com/47539414/70962390-0ebf1680-2053-11ea-8f74-024ec2b68415.png)

### Steps to reproduce:

No special steps, memory leak appears to be related to procstat input plugin:
![image](https://user-images.githubusercontent.com/47539414/70962479-68274580-2053-11ea-8700-ed370071d366.png)

![image](https://user-images.githubusercontent.com/47539414/70962498-74ab9e00-2053-11ea-9738-de0e86f6a806.png)

**Comment by danielnelson** [comment]
It looks like there are as many frees 16MB as allocs 16.1MB.  What is the query behind `procstat.mean (process_name: telegraf.exe)`?

**Comment by mafinn** [comment]
![image](https://user-images.githubusercontent.com/47539414/71000889-11008f80-20aa-11ea-9c7a-a32e6df8fa66.png)

![image](https://user-images.githubusercontent.com/47539414/71001367-2aeea200-20ab-11ea-9702-9207fe700992.png)

**Comment by mafinn** [comment]
Eventually, **telegraf** will consume enough memory that it will be stopped:
![image](https://user-images.githubusercontent.com/47539414/71001594-946eb080-20ab-11ea-9cb5-0fe19ecde152.png)

**Comment by mafinn** [comment]
It's interesting to see the change in slope of memory growth before and after the restart at ~10am. I modified the config from `exe = ".*"` to 7 instances of procstat, each calling a unique process name. This also decreased the gather time from 17 seconds to 3 seconds.

![image](https://user-images.githubusercontent.com/47539414/71044038-86964b00-20fe-11ea-9d4c-f938682a04c5.png)

**Comment by danielnelson** [comment]
I've been running this configuration for about 4 hours and I'm not seeing this pattern, however I don't have any real load on the system.

Could you add the `--pprof-addr=:6060` option when starting Telegraf and after the process RSS doubles from startup, go to `http://localhost:6060/debug/pprof/heap` and attach this file?

**Comment by mafinn** [comment]
@danielnelson can do. Can I send it directly to you?

**Comment by danielnelson** [comment]
Yes, email address is on my profile page.

**Comment by danielnelson** [comment]
Can you show a couple hours of `internal_memstats sys_bytes`?

**Comment by mafinn** [comment]
Mostly fixed at 19.81 MB

24 hours:
![image](https://user-images.githubusercontent.com/47539414/71118667-8b163e80-21a6-11ea-96d3-c59592869dbc.png)

3 hours:
![image](https://user-images.githubusercontent.com/47539414/71118691-97020080-21a6-11ea-9dc5-8b36ac79817d.png)

**Comment by danielnelson** [comment]
The RSS is rising but Go doesn't seem to know about it, I didn't see any interesting objects in the memory profiles either.  Just thinking aloud but could it be lost to Go or maybe it is leaked in a dll call, I'm not sure.  I'm unable to replicate as well even when spawning new processes on the system.

Do you know if this is a new issue with Telegraf 1.13?  Could you compare against [1.12.6][] and [1.11.5][]?

[1.12.6]: https://github.com/influxdata/telegraf/releases/tag/1.12.6
[1.11.5]: https://github.com/influxdata/telegraf/releases/tag/1.11.5

**Comment by mafinn** [comment]
![image](https://user-images.githubusercontent.com/47539414/71213434-82df0180-2281-11ea-91b5-319c00ea82e6.png)

**Comment by mafinn** [comment]
![image](https://user-images.githubusercontent.com/47539414/71214914-61cbe000-2284-11ea-8327-f25e6e44d257.png)

**Comment by danielnelson** [comment]
For the record, here are my numbers with Telegraf 1.13.0 (Windows 7):

![2019-12-19-144203_1370x918_scrot](https://user-images.githubusercontent.com/1048079/71215930-ea8b5180-226d-11ea-95f7-2d6e2e677c1e.png)

At first it trends up, but levels off around 55MB RSS.  It seems to take about 3 hours before it maxes.

**Comment by mafinn** [comment]
So far, look across our different instances of windows, I've only seen this occur on windows server 2016.

**Comment by danielnelson** [comment]
Related? https://github.com/go-ole/go-ole/issues/135

**Comment by mafinn** [comment]
2016:
![image](https://user-images.githubusercontent.com/47539414/71217881-5204c980-228d-11ea-85dc-f4e1aab339b0.png)

2012:
![image](https://user-images.githubusercontent.com/47539414/71217903-7496e280-228d-11ea-96ad-5cf5b0ceac78.png)

**Comment by danielnelson** [comment]
What should I be looking at here?

**Comment by mafinn** [comment]
Just trying to find a way to capture the difference in DLL's in use between 2016 and not-2016.

**Comment by danielnelson** [comment]
I'm going to try installing WMF 5.1 to see if it causes the error as suggested in https://github.com/go-ole/go-ole/issues/135#issuecomment-283440299.

**Comment by mafinn** [comment]
Quick update, I installed WMF 5.1 on a Windows 2012 (non-R2) box, and it didn't cause the memory leak.

**Comment by danielnelson** [comment]
Same on my Windows 7 system, WMF 5.1 had no effect.

**Comment by danielnelson** [comment]
Still not reproducing the leak with a Windows 10 Pro VM:

![2020-01-08-220402_1332x874_scrot](https://user-images.githubusercontent.com/1048079/72042297-ec3fa800-3262-11ea-97c5-844691a1822e.png)

**Comment by danielnelson** [comment]
I have been able to reproduce this on a Windows 2016 VM running in Azure.  Will update if I can find a way to reduce or eliminate the leaked memory.

**Comment by romanblachman** [comment]
This still continues to be an issue with the latest Windows Server 2016 - WMI based metrics cause memory leak. We have been using WMI queries and are able to bypass this by calling CoInitializeEx only once per thread, but it seems that telegraf leaks based on the number of WMI metrics, probably since CoInitializeEx is called for every query?

![image](https://user-images.githubusercontent.com/5787391/75145425-98e5c400-5701-11ea-89a3-0360e8016b7a.png)
In the attached screenshot, the win_proc for telegraf (Working Set Private metric) increases to 100MB with-in 7 days.

**Comment by romanblachman** [comment]
@danielnelson how does win_proc access WMI metrics? We have fixed the issue by calling CoInitializeEx once per thread to avoid the leak from happening when using WMI metrics in our code, instead of calling it once per WMI query.

**Comment by romanblachman** [comment]
@danielnelson Looks like datadog had the same issue with their WMI sampler https://github.com/DataDog/integrations-core/pull/3987 which tells clearly that this issue is with Windows 2016 memory leak when calling CoInitalize for each WMI query.

After reviewing the code for telegraf it seems like you rely on win_pdh library that does the actual Win32 calls, and I couldn't find the call to the CoInitalize so I'm not sure how to help.

**Comment by Jaeyo** [comment]
any update?

**Comment by danielnelson** [comment]
Can you retest with this build [telegraf-1.15.0~d78dfac1_windows_amd64.zip](https://56669-33258973-gh.circle-artifacts.com/0/build/windows/amd64/telegraf-1.15.0~d78dfac1_windows_amd64.zip)?

**Comment by romanblachman** [comment]
Will do!

**Comment by romanblachman** [comment]
Looks like this has been resolved, the WMI leak is gone on Windows Server 2016!

48 hours running with **1.12.6**:

![image](https://user-images.githubusercontent.com/5787391/83971318-832c9100-a8e3-11ea-92c4-f9095c204485.png)

24 hours running with **1.15.0**:

![image](https://user-images.githubusercontent.com/5787391/83971342-9b041500-a8e3-11ea-9648-c66d24af180c.png)

thank you @danielnelson 

what's the eta on 1.15.0?

**Comment by danielnelson** [comment]
Great news, thanks for testing.

I expect 1.15.0 to be released sometime in the first half of July.
